File: ChainOfResponsibility/ChainOfResATM.py
from abc import ABC, abstractmethod

class ATMHandlerAbstract(ABC):
    def __init__(self, successor, denomination):
        self._successor = successor
        self._denomination = denomination

    @abstractmethod
    def handle_request(self, amount):
        pass

class ATMHandler(ATMHandlerAbstract):
    def handle_request(self, amount):
        q = int(amount // self._denomination)
        r = amount % self._denomination
        if q != 0:
            print("%s notes of %s"%(q,self._denomination))

        if r != 0 and self._successor is not None:
            self._successor.handle_request(r)


# Client Code
ch = ATMHandler(ATMHandler(ATMHandler(ATMHandler(None, 10), 50), 100), 1000)
ch.handle_request(5560)

"""
5 notes of 1000
5 notes of 100
1 notes of 50
1 notes of 10
"""

File: ChainOfResponsibility/ChainOfResATM.java
abstract class ATMHandlerAbstract {
    protected ATMHandlerAbstract successor;
    protected int denomination;

    public ATMHandlerAbstract(ATMHandlerAbstract successor, int denomination) {
        this.successor = successor;
        this.denomination = denomination;
    }

    public abstract void handleRequest(int amount);
}

class ATMHandler extends ATMHandlerAbstract {
    public ATMHandler(ATMHandlerAbstract successor, int denomination) {
        super(successor, denomination);
    }

    @Override
    public void handleRequest(int amount) {
        int q = amount / denomination;
        int r = amount % denomination;

        if (q != 0) {
            System.out.println(q + " notes of " + denomination);
        }

        if (r != 0 && successor != null) {
            successor.handleRequest(r);
        }
    }
}

public class ChainOfResATM {
    public static void main(String[] args) {
        ATMHandlerAbstract handler = new ATMHandler(
                new ATMHandler(
                        new ATMHandler(
                                new ATMHandler(null, 10), 50
                        ), 100
                ), 1000
        );

        handler.handleRequest(5560);
    }
}

/*
5 notes of 1000
5 notes of 100
1 notes of 50
1 notes of 10
 */